examples/hhl.py

```diff
@@ -0,0 +1,414 @@
+
+|x> = A**-1 |b> / || A**-1 |b> ||
+
+The algorithm uses 3 sets of qubits: a single ancilla qubit, a register (to
+store eigenvalues of A), and memory qubits (to store |b> and |x>). The
+following are performed in order:
+1) Quantum phase estimation to extract eigenvalues of A
+2) Controlled rotations of ancilla qubit
+3) Uncomputation with inverse quantum phase estimation
+
+For details about the algorithm, please refer to the paper in the
+REFERENCE section below. The following description uses variables defined
+in the paper.
+
+This example is an implementation of the HHL algorithm for arbitrary 2x2
+Hermitian matrices. The output of the algorithm is a display of Pauli
+observables of |x>. Note that the accuracy of the result depends on the
+following factors:
+* Register size
+* Choice of parameters C and t
+
+The result is perfect if
+* Each eigenvalue of the matrix is in the form
+
+  2π/t * k/N,
+
+  where 0≤k<N, and N=2^n, where n is the register size. In other words, k is a
+  value that can be represented exactly by the register.
+* C ≤ 2π/t * 1/N, the smallest eigenvalue that can be stored in the circuit.
+
+The result is good if the register size is large enough such that for every
+pair of eigenvalues, the ratio can be approximated by a pair of possible
+register values. Let s be the scaling factor from possible register values to
+eigenvalues. One way to set t is
+
+t = 2π/sN
+
+For arbitrary matrices, because properties of their eigenvalues are typically
+unknown, parameters C and t are fine-tuned based on their condition number.
+
+
+=== REFERENCE ===
+Coles, Eidenbenz et al. Quantum Algorithm Implementations for Beginners
+https://arxiv.org/abs/1804.03719
+
+=== CIRCUIT ===
+Example of circuit with 2 register qubits.
+
+(0, 0): ─────────────────────────Ry(θ₄)─Ry(θ₁)─Ry(θ₂)─Ry(θ₃)──────────────M──
+                     ┌──────┐    │      │      │      │ ┌───┐
+(1, 0): ─H─@─────────│      │──X─@──────@────X─@──────@─│   │─────────@─H────
+           │         │QFT^-1│    │      │      │      │ │QFT│         │
+(2, 0): ─H─┼─────@───│      │──X─@────X─@────X─@────X─@─│   │─@───────┼─H────
+           │     │   └──────┘                           └───┘ │       │
+(3, 0): ───e^iAt─e^2iAt───────────────────────────────────────e^-2iAt─e^-iAt─
+
+Note: QFT in the above diagram omits swaps, which are included implicitly by
+reversing qubit order for phase kickbacks.
+"""
+
+import math
+import cirq
+import numpy as np
+
+
+class PhaseEstimation(cirq.MultiQubitGate):
+    """
+    A gate for Quantum Phase Estimation.
+
+    U is the unitary matrix whose phases will be estimated.
+    The last qubit stores the eigenvector; all other qubits store the
+    estimated phase, in big-endian.
+    """
+
+    def __init__(self, num_qubits, U, invert=False):
+        super().__init__(num_qubits)
+        self.U = U
+        self.invert = invert
+
+    def _decompose_(self, qubits):
+        if self.invert:
```

Comment:
Once https://github.com/quantumlib/Cirq/pull/1407 is merged, you can remove this inverse stuff.